lib/cassandra.rb

```diff
@@ -219,15 +235,77 @@ def self.cluster(options = {})
   # @return [Cassandra::Future<Cassandra::Cluster>] a future resolving to the
   #   cluster instance.
   def self.cluster_async(options = {})
-    options = options.select do |key, value|
-      [ :credentials, :auth_provider, :compression, :hosts, :logger, :port,
-        :load_balancing_policy, :reconnection_policy, :retry_policy, :listeners,
-        :consistency, :trace, :page_size, :compressor, :username, :password,
-        :ssl, :server_cert, :client_cert, :private_key, :passphrase,
-        :connect_timeout, :futures_factory, :datacenter, :address_resolution,
-        :address_resolution_policy, :idle_timeout, :heartbeat_interval, :timeout,
-        :synchronize_schema, :schema_refresh_delay, :schema_refresh_timeout,
-        :shuffle_replicas, :client_timestamps
+    options = validate_and_massage_options(options)
+    hosts = []
+
+    Array(options.fetch(:hosts, '127.0.0.1')).each do |host|
+      case host
+      when ::IPAddr
+        hosts << host
+      when ::String # ip address or hostname
+        Resolv.each_address(host) do |ip|
+          hosts << ::IPAddr.new(ip)
+        end
+      else
+        raise ::ArgumentError, ":hosts must be String or IPAddr, #{host.inspect} given"
+      end
+    end
+
+    if hosts.empty?
+      raise ::ArgumentError, ":hosts #{options[:hosts].inspect} could not be resolved to any ip address"
+    end
+
+    hosts.shuffle!
+  rescue => e
+    futures = options.fetch(:futures_factory) { return Future::Error.new(e) }
+    futures.error(e)
+  else
+    driver = Driver.new(options)
+    driver.connect(hosts)
+  end
+
+  # @private
+  def self.validate_and_massage_options(options)
```

Comment:
this should probably be called `filter_options` or `clean_options` since neither validation nor massaging is happening on the hash